qiskit/tools/parser.py

```diff
@@ -0,0 +1,149 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2019.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+# pylint: disable=invalid-name
+
+"""Helper function to parse string expression given by backends."""
+
+import ast
+import operator
+import re
+
+import cmath
+
+from qiskit.exceptions import QiskitError
+
+
+# valid functions
+_math_ops = {key: value for (key, value) in vars(cmath).items() if key[0] != '_'}
+
+# valid parameters
+_param_regex = r'(P\d+)'
+
+
+def _eval_expr(parsed_expr, locals_dict):
+    """ Evaluate expression without python built-in eval function.
+
+    Args:
+        parsed_expr (ast.Expression): Parsed expression to evaluate.
+        locals_dict (dict): Dictionary of user arguments.
+
+    Returns:
+        complex: Evaluated value with given arguments.
+
+    Raises:
+        QiskitError: If expression is not in proper format.
+    """
+
+    def _visit_node(node):
+        if isinstance(node, ast.UnaryOp):
+            # Unary operation
+            operand = _visit_node(node.operand)
+            if isinstance(node.op, ast.UAdd):
+                opr = operator.add
+            elif isinstance(node.op, ast.USub):
+                opr = operator.sub
+            else:
+                raise QiskitError('Operator %s is not supported.' % node.op)
+            return opr(0, operand)
+        elif isinstance(node, ast.BinOp):
+            # Binary operation
+            l_value = _visit_node(node.left)
+            r_value = _visit_node(node.right)
+            if isinstance(node.op, ast.Add):
+                opr = operator.add
+            elif isinstance(node.op, ast.Sub):
+                opr = operator.sub
+            elif isinstance(node.op, ast.Mult):
+                opr = operator.mul
+            elif isinstance(node.op, ast.Div):
+                opr = operator.truediv
+            elif isinstance(node.op, ast.Pow):
+                opr = operator.pow
+            else:
+                raise QiskitError('Operator %s is not supported.' % node.op)
+            return opr(l_value, r_value)
+        elif isinstance(node, ast.Num):
+            # number
+            try:
+                return complex(node.n)
```

Comment:
will this cause every evaluated expression to return a complex, even if the input number was a float?